services/mac_manager/src/test/java/com/futurewei/alcor/macmanager/controller/MacControllerTest.java

```diff
@@ -113,43 +77,69 @@ public void test_createMacState() throws Exception {
         MacStateJson macStateJson = new MacStateJson(macState);
         ObjectMapper objectMapper = new ObjectMapper();
         String json = objectMapper.writeValueAsString(macStateJson);
+        this.mockMvc.perform(post("/macs")
+                .content(json)
+                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON))
+                .andDo(print())
+                .andExpect(status().isCreated());
+    }
 
+    @Test
+    public void test_createMacState_invalidInput_null() throws Exception {
+        MacState macState = new MacState("", "project3", "vpc3", "port3", "Active");
+        MacStateJson macStateJson = null;
+        ObjectMapper objectMapper = new ObjectMapper();
+        String json = objectMapper.writeValueAsString(macStateJson);
         this.mockMvc.perform(post("/macs")
                 .content(json)
                 .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON))
-                .andExpect(status().isCreated())
-                .andDo(print());
+                .andDo(print())
+                .andExpect(status().isBadRequest());
     }
 
     @Test
     public void test_getMacStateByMacAddress() throws Exception {
-        MacState macState = new MacState("", "project1", "vpc1", "port2", "Active");
-        String strTestMac = createMacState(macState);
+        MacState macState = new MacState("AA-BB-CC-01-01-01", "project1", "vpc1", "port2", "Active");
+        MacStateJson macStateJson = new MacStateJson(macState);
+        String strTestMac = "AA-BB-CC-01-01-01";
+        Mockito.when(mockController.getMacStateByMacAddress(strTestMac)).thenReturn(macStateJson);
```

Comment:
Glad to see the UTs are finally working.

Reply:
Currently the UTs mocks from the web level, which means that the service and db access codes are untested actually. We should only mock low lever db access, instead of web layer on the top.

As discussed on the Slack channel on May 12th @alcor channel, our microservice UTs should fall into one of three categories:
(1) Controller/Service UTs, which only use Mockito but without using MockIgniteService
(2) Repo UTs which uses MockIgniteService
(3) Microservice-level 'e2e' UTs , to cover controller/Service/repo with MockIgniteService except intra-service communication.

An example to follow: https://github.com/futurewei-cloud/alcor/blob/master/services/port_manager/src/test/java/com/futurewei/alcor/portmanager/controller/PortControllerTest.java

Feel free to ping me if you have any questions or get some clarification.